optimization/ht_functions.py

```diff
@@ -0,0 +1,385 @@
+        self.D_e = 2.0 * self.r_channel
+    
+    def characterize_flow(self):
+        """Calculate important non-dim and dim flow parameters. These parameters
+        are required to determine generation per fuel channel.
+        
+        Modified Attributes:
+        --------------------
+            G_dot: mass flux through all flow channels [kg/m^2-s]
+            v: flow velocity [m/s]
+            Re: Reynolds number [-]
+            Pr: Prandtl number [-]
+            Nu: Nusselt number [-]
+            f: friction factor [-]
+            h_bar: heat transfer coefficient [W/m^2-K]
+        """
+        # calculate mass flux
+        self.G_dot = m_dot / (self.A_flow * self.guess)
+        # calculate flow velocity from mass flux
+        self.v = self.G_dot / rho_cool
+        # calculate Reynolds Number
+        self.Re = rho_cool * self.v * self.D_e / mu
+        # calculate Prandtl number
+        self.Pr = Cp_cool * mu / k_cool
+        # nusselt correlation is Eq (9-22) from El-Wakil Nuclear Heat Transport
+        self.Nu = 0.023*math.pow(self.Re,0.8)*math.pow(self.Pr, 0.4)
+        # friction factor for pressure drop correlation El Wakil (9-4)
+        self.f = 0.184 / math.pow(self.Re, 0.2)
+        # heat transfer coefficient 
+        self.h_bar = self.Nu * k_cool / self.D_e
+
+    def get_q_per_channel(self):
+        """Calculate achievable average volumetric generation:
+        This method uses previously set geometry and calculated flow
+        parameters to determine the maximum-achievable volumetric generation
+        for each fuel channel. Calculates number of fuel channels required
+        for desired thermal output.
+        
+        Modified Attributes:
+        --------------------
+            R_fuel: resistance term (conduction in the fuel) [W/K]
+            R_clad: resistance term (conduction in the clad) [W/K]
+            R_conv: resistance term (convection to the fluid) [W/K]
+            R_tot: total resistance to HT [W/K]
+            q_per_channel: total generation in fuel channel [W]
+            q_bar: axially-averaged volumetric generation in fuel [W]
+            N_channels: required channels for desired Q [-]
+        """
+        r_i = self.r_channel + self.c
+        r_o = self.pitch / math.sqrt(3)
+       
+        # El Wakil (9-62) calculates max q''' at axial centerline
+        self.R_fuel = (r_o**2 / (4*k_fuel)) * ((r_i/r_o)**2 - 2*math.log(r_i/r_o) - 1)
+        
+        self.R_clad = (r_o**2)/2 * (1-(r_i/r_o)**2)*\
+        math.log(r_i/(r_i-self.c)) / k_clad
+        
+        self.R_conv = (r_o**2)/2 * (1-(r_i/r_o)**2)*\
+        1 / (self.h_bar*(r_i - self.c))
+        
+        self.R_tot = self.R_fuel + self.R_clad + self.R_conv
+        
+        # calculate centerline volumetric generation
+        q_trip_max = self.dt / self.R_tot
+        
+        # consider axial flux variation
+        self.q_per_channel = 2 * q_trip_max * self.A_fuel * self.L / math.pi
```

Comment:
Not sure this is the right way to calculate `q_per_channel` given your annular approximation.  Perhaps the `q_per_channel` should be calculated with the annular fuel cross-section area, and not this hexagonal one.

Reply:
I debated over this for a while, q_trip_max is a conservative estimate for q''' using an annular approximation. After that, I need to average this value over the axial length of the channel. I wanted to use the actual geometry in this calculation. Esp, since the annular approx. over-estimates the A_fuel / channel.